concordium-consensus/src/Concordium/KonsensusV1/TreeState/Implementation.hs

```diff
@@ -0,0 +1,737 @@
+        Just sb -> return . Just =<< mkBlockPointer sb
+
+-- |Turn a 'PendingBlock' into a live block.
+-- This marks the block as 'MemBlockAlive' in the block table, records the arrive time of the block,
+-- and returns the resulting 'BlockPointer'.
+-- The hash of the block state MUST match the block state hash of the block; this is not checked.
+-- [Note: this does not affect the branches.]
+makeLiveBlock :: (MonadState (SkovData pv) m) => PendingBlock -> PBS.HashedPersistentBlockState pv -> BlockHeight -> UTCTime -> m (BlockPointer pv)
+makeLiveBlock pb st height arriveTime = do
+    let bp =
+            BlockPointer
+                { bpInfo = BlockMetadata{bmReceiveTime = pbReceiveTime pb, bmArriveTime = arriveTime, bmHeight = height},
+                  bpBlock = NormalBlock (pbBlock pb),
+                  bpState = st
+                }
+    blockTable . liveMap . at' (getHash pb) ?=! MemBlockAlive bp
+    return bp
+
+-- |Marks a block as dead.
+-- This expunges the block from memory
+-- and registers the block in the dead cache.
+-- [Note: this does not affect the branches.]
+markBlockDead :: (MonadState (SkovData pv) m) => BlockHash -> m ()
+markBlockDead blockHash = do
+    blockTable . liveMap . at' blockHash .=! Nothing
+    blockTable . deadBlocks %=! insertDeadCache blockHash
+
+-- |Mark a live block as dead. In addition, purge the block state and maintain invariants in the
+-- transaction table by purging all transaction outcomes that refer to this block.
+-- [Note: this does not affect the branches.]
+markLiveBlockDead ::
+    ( MonadState (SkovData pv) m,
+      BlockStateStorage m,
+      GSTypes.BlockState m ~ PBS.HashedPersistentBlockState pv
+    ) =>
+    BlockPointer pv ->
+    m ()
+markLiveBlockDead bp = do
+    let bh = getHash bp
+    markBlockDead bh
+    purgeBlockState $ bpState bp
+    mapM_ (markTransactionDead bh) (blockTransactions bp)
+
+-- |Mark a block as pending in the block table.
+-- (Note, this does not update the pending block table.)
+markPending :: (MonadState (SkovData pv) m) => PendingBlock -> m ()
+markPending pb = blockTable . liveMap . at' (getHash pb) ?=! MemBlockPending pb
+
+-- * Operations on pending blocks
+
+-- $pendingBlocks
+-- Pending blocks are conceptually stored in a min priority search queue,
+-- where multiple blocks may have the same key, which is their parent,
+-- and the priority is the block's round number.
+-- When a block arrives (possibly dead), its pending children are removed
+-- from the queue and handled.  This uses 'takePendingChildren'.
+-- When a block is finalized, all pending blocks with a lower or equal round
+-- number can be handled (they will become dead, since they can no longer
+-- join the tree).  This uses 'takeNextPendingUntil'.
+
+-- |Add a block to the pending block table and queue.
+-- (Note, this does not update the block table.)
+addPendingBlock :: (MonadState s m, HasPendingBlocks s) => PendingBlock -> m ()
+addPendingBlock pb = do
```

Comment:
By looking at the type definition(s) I believe it's as strict as it can be. But if I'm wrong please say so.

Reply:
I mean this depends on uses of this function. If the pending block is evaluated fully before this is called (which it probably is since we verify the transactions and look at signatures, etc) then it doesn't matter. It is not the pendingBlock definition that worries me, it is that `pb` here is just added to a data structure without being inspected. 

But it is probably fine.

Reply:
I see. I can bang it for good measurement. But it is probably fine. But just to be strictly safe so to speak..